code/filesystem/locator/abstract.php

```diff
@@ -0,0 +1,241 @@
+        }
+
+        return $this->__location_cache[$url];
+    }
+
+    /**
+     * Parse the url
+     *
+     * @param  string $url The language url
+     * @return array
+     */
+    public function parseUrl($url)
+    {
+        $scheme   = parse_url($url, PHP_URL_SCHEME);
+        $domain   = parse_url($url, PHP_URL_HOST);
+        $basename = pathinfo($url, PATHINFO_BASENAME);
+
+        $parts = explode('.', $basename);
+
+        if(count($parts) == 3)
+        {
+            $type  =  array_pop($parts);
+            $format = array_pop($parts);
+            $file   = array_pop($parts);
+        }
+        else
+        {
+            $type   = '*';
+            $format = array_pop($parts);
+            $file   = array_pop($parts);
+        }
+
+        $path = str_replace(array($scheme.'://', $scheme.':'), '', dirname($url));
+        $parts = explode('/', $path);
+
+        $info = array(
+            'type'    => $type,
+            'domain'  => $domain,
+            'package' => array_shift($parts),
+            'path'    => implode('/', $parts),
+            'file'    => $file,
+            'format'  => $format ?: '*',
+        );
+
+        return $info;
+    }
+
+    /**
+     * Register a path template
+     *
+     * @param  string $template   The path template
+     * @param  bool $prepend      If true, the template will be prepended instead of appended.
+     * @return FilesystemLocatorAbstract
+     */
+    public function registerPathTemplate($template, $prepend = false)
+    {
+        if($prepend) {
+            array_unshift($this->__path_templates, $template);
+        } else {
+            array_push($this->__path_templates, $template);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Get the list of path templates
+     *
+     * @param  string $url   The Template url
+     * @return array The path templates
+     */
+    public function getPathTemplates($url)
+    {
+        return $this->__path_templates;
+    }
+
+    /**
+     * Get a path from an file
+     *
+     * Function will check if the path is an alias and return the real file path
+     *
+     * @param  string $file The file path
+     * @return string The real file path
+     */
+    final public function realPath($file)
+    {
+        $result = false;
+        $path   = dirname($file);
+
+        // Is the path based on a stream?
+        if (strpos($path, '://') === false)
+        {
+            // Not a stream, so do a realpath() to avoid directory traversal attempts on the local file system.
+            $path = realpath($path); // needed for substr() later
+            $file = realpath($file);
+        }
+
+        // The substr() check added to make sure that the realpath() results in a directory registered so that
+        // non-registered directories are not accessible via directory traversal attempts.
+        if (file_exists($file) && substr($file, 0, strlen($path)) == $path) {
```

Comment:
You could make this a ===. The result would be the same dirname expects and returns string and so does substr. Or am I missing something ?